shell/browser/ui/cocoa/electron_ns_window.mm

```diff
@@ -224,6 +224,18 @@ - (void)toggleFullScreenMode:(id)sender {
     shell_->SetSimpleFullScreen(!is_simple_fs);
   } else {
     bool maximizable = shell_->IsMaximizable();
+
+    // [NSWindow -toggleFullScreen] is an asynchronous operation, which means
+    // that it's possible to call it while a fullscreen transition is currently
+    // in process. This can create weird behavior (incl. phantom windows),
+    // so we want to block the operation until any current transitions have
+    // completed.
+    while (shell_->entering_fullscreen() || shell_->exiting_fullscreen()) {
+      [[NSRunLoop currentRunLoop]
+             runMode:NSDefaultRunLoopMode
+          beforeDate:[NSDate dateWithTimeIntervalSinceNow:.5]];
```

Comment:
🚨 

Are we sure we want to run a nested run loop? I think this can, for example, re-enter JS. Which probably isn't what we want.

Also, using a specific time interval seems very flaky to me. Is this the same time interval against which the animation is measured? How will we know if macOS tweaks the length of the animation? If there's high CPU load, does the animation take longer? We should use some kind of event to know when this is done, instead of a time interval.

Reply:
Refactored to schedule transitions instead!